instrumentation/opentelemetry-instrumentation-urllib3/src/opentelemetry/instrumentation/urllib3/__init__.py

```diff
@@ -170,26 +172,84 @@ def instrumented_urlopen(wrapped, instance, args, kwargs):
             SpanAttributes.HTTP_URL: url,
         }
 
+        links = None
+        # If current call is a retry call
+        if retry and len(retry.history):
+            # History reflects how many retry attempts have been made
+            # Only add retry count on retry scenarios
+            if not is_redirect(retry.history[-1].status):
+                span_attributes["http.retry_count"] = len(retry.history)
+            prev_span_context = get_current_span().get_span_context()
+            if prev_span_context is INVALID_SPAN_CONTEXT:
+                # If current span is root but we are in a retry scenario, this
+                # is a redirect scenario
+                prev_span_context = getattr(retry, "_OT_retry_prev_span_context", None)
+            if prev_span_context is not INVALID_SPAN_CONTEXT:
+                links = (Link(prev_span_context),)
+
         with tracer.start_as_current_span(
-            span_name, kind=SpanKind.CLIENT, attributes=span_attributes
+            span_name, kind=SpanKind.CLIENT, attributes=span_attributes, links=links,
```

Comment:
won't every retry span become a CLIENT span in this case? This would be fine if all the client spans were siblings but the PR description suggests in some cases the retry spans can be children of the previous retry attempts. In that case having a CLIENT span being a child of another CLIENT span could be a problem.

Reply:
That is a good point. I'd be open to setting the children retry attempts to `INTERNAL` maybe?

Reply:
That would be better than nested CLIENT spans but still not ideal. Ideally a CLIENT span should have a SERVER span as a child but not sure if Otel mandates that. Taking a step back, I feel the a better model might be to always have one INTERNAL span representing the overall request and then have each attempt as a child with kind set to CLIENT.  For example, in case where the request succeeds in the first attempt the trace would look like:


```
request (internal)
     |
     -- GET 200 (client)
         |
         --- server (SERVER)
```

and multiple attempts could look like:


```
request (internal)
     |
     -- GET / - 500 (client)
     |
     -- GET / - 500 (client)
     |
     -- GET / - 200 (client)
         |
         --- server (SERVER)
```

and a redirect could include server span/response as well



```
request (internal)
     |
     -- GET / - 302 (client)
     |
     -- GET /hello - 200 (client)
         |
         --- server (SERVER)
```

Reply:
Interesting. It looks like the dot net prototype is using CLIENT [spans](https://github.com/open-telemetry/opentelemetry-dotnet/pull/2756/files) for each retry. Each span created from a retry/redirect is essentially a duplicate of the original request span (besides from the attributes/link defined in the spec). @denisivan0v any input on this parent/child relationship with these span kinds?

Reply:
> I feel the a better model might be to always have one INTERNAL span representing the overall request and then have each attempt as a child with kind set to CLIENT.

This is the valid point. There was a similar discussion within Instrumentation SIG https://github.com/open-telemetry/opentelemetry-specification/pull/2078#discussion_r755521314. Essentially, we have 4 options here:

- retry/redirect span being a child
- retry/redirect span created for a "physical"/CLIENT request only
- retry/redirect span created for a "logical"/INTERNAL request only
- retry/redirect span created for both "physical" and "logical" (+ adding the "logical" span in case there was a retry/redirect)

The idea of having spans for "physical" requests only + adding links to previous try/redirect span allows us to model it in a simplest way and avoid duplication - there will be no additional "logical"/INTERNAL span if case the initial request finished with 2xx (which is backward-compatible with the current behaivor). That said, it seems spans for "physical" requests _should_ be created as siblings to keep request-response semantics.

I'm in progress of adding these clarifications to the https://github.com/open-telemetry/opentelemetry-specification/pull/2078, please let me know if sounds reasonable to you.

The [PR for .NET](https://github.com/open-telemetry/opentelemetry-dotnet/pull/2756/files) implements the mentioned proposal (so no duplicate of the original request span).

Reply:
From understanding, this proposal means that spans created from retry/redirect will have no parent/child relationship with each other or the original request, but will be LINKed together. @owais Thoughts?

Reply:
> From understanding, this proposal means that spans created from retry/redirect will have no parent/child relationship with each other or the original request, but will be LINKed together. @owais Thoughts?

That makes sense to me. Just to clarify, they'll still all be children of whatever the active span was before the first request attempt was made, right? What if there was no parent? Would we end up with N number of traces with N being number of attempts? May be we should create an encompassing span representing the whole operation to overcome this?

Reply:
> Would we end up with N number of traces with N being number of attempts

I think this would indeed be the case. Is there a huge need for this? I can foresee the implementation of that being kind of ugly, especially because we don't know that it is a retry/redirect scenario until after the second call is made.

Reply:
Imagine an auto-instrumented script or cronjob that is invoked and first (and possibly only) script it does is make an HTTP call to an external service. In case of retries/redirects, we'll end up with multiple traces with linked spans between them while as logically it was a single operation and should be a single trace.

Reply:
Would the encompassing span be of type INTERNAL?

Reply:
That is a good point. Personally I'm not fond of the idea of "automatically" creating spans that don't represent operations, it makes the already "magic-like" implementation have even more hidden behaviour. @denisivan0v has brought up some good points about how maybe an encompassing span is not totally necessary. Perhaps he can chime in more in this discussion. I was wondering if this was behaviour we had to decide/define now, or perhaps we can have the first iteration of this be a simple implementation and we can iterate on it later so as to push the spec along?

Reply:
I think we can totally iterate and leave these cases out as long as the feedback about these cases bubbles up to the spec level and it is addressed there eventually.

Reply:
Many thanks for this discussion, it helped to understand what should be more clarified in [the PR to the spec](https://github.com/open-telemetry/opentelemetry-specification/pull/2078).
I've re-worded the whole section to address these points, please have a look here https://github.com/open-telemetry/opentelemetry-specification/blob/bb265215d6b89dc0cfdae37f41123d6bb7986f9e/specification/trace/semantic_conventions/http.md#http-request-retries-and-redirects.

As for the encompassing span representing the whole operation, I've put it as 
```
No span is created corresponding to the "logical" (encompassing) request.
```
since it was a consensus within Instrumentation SIG folks to keep it as simple as possible and avoid any kind of hidden and automatic span creation.

Reply:
Nice! This makes sense to me. I have made the proposed changes as part of this PR.